frontend/src/main/scala/bloop/io/SourceWatcher.scala

```diff
@@ -61,22 +63,38 @@ final class SourceWatcher(project: Project, dirs0: Seq[Path], logger: Logger) {
       }
     )
 
-    val watchingTask = Task {
+    // Use Java's completable future because we can stop/complete it from the cancelable
+    val watcherHandle = watcher.watchAsync(ExecutionContext.ioExecutor)
+    val watchController = Task {
       logger.info(s"File watching $dirsCount directories...")
-      try watcher.watch()
+      try watcherHandle.get()
       finally watcher.close()
-    }.doOnCancel(Task {
+      logger.debug("The file watcher was successfully closed.")
+    }
+
+    val watchCancellation = Cancelable { () =>
+      watchingEnabled = false
+      watcherHandle.complete(null)
```

Comment:
Can the `null` find its way to the `fileEventConsumer`?

Reply:
Yeah, this works, Java expects `Void` and for that you need to use null in Scala.